pkg/util/retryableerror/retryableerror.go

```diff
@@ -0,0 +1,59 @@
+package retryableerror
+
+import (
+	"time"
+
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+)
+
+// Error represents an error for an operation that should be retried after the
+// specified duration.
+type Error interface {
+	error
+	// After is the time period after which the operation that caused the
+	// error should be retried.
+	After() time.Duration
+}
+
+// New returns a new RetryableError with the given error and time period.
+func New(err error, after time.Duration) Error {
+	return retryableError{err, after}
+}
+
+type retryableError struct {
+	error
+	after time.Duration
+}
+
+// After returns the time period after which the operation that caused the error
+// should be retried.
+func (r retryableError) After() time.Duration {
+	return r.after
+}
+
+// NewMaybeRetryableAggregate converts a slice of errors, some of which may be
+// RetryableError values, into a RetryableAggregate interface, which is itself
+// an implementation of the error interface.  Nil values will be filtered from
+// the slice.  If the filtered slice is empty, the return value will be nil.  If
+// any values are RetryableError values, the result will be a RetryableError,
+// with RetryAfter equal to the minimum of all the RetryableError values'
+// RetryAfter values.
+func NewMaybeRetryableAggregate(errs []error) error {
```

Comment:
Not 100% sure about these semantics... the way controller-runtime works is that if Reconcile() returns an error, the request is unconditionally re-queued with the default rate limiting stuff. Only if a nil err is returned will RetryAfter be honored. So if an aggregate contains a mix of types, we're inverting the usual behavior in response to non-retryable errors... Not sure if it matters, but it seemed weird enough to at least bring up.

Reply:
On the other hand, this is far more principled an approach than we currently have where literally all errors are equal

Reply:
`NewMaybeRetryableAggregate` could set `retryAfter` to 0 if `errs` contains a non-retryable error.  However, controller-runtime takes `RequeueAfter: 0` to mean "don't requeue"; would it be too weird to have different semantics (where having retryable error type means retry, and if `After()` is 0, then retry immediately)?

Reply:
Maybe it would be less weird just to return a retryable error only if `errs` is all retryable errors.

Reply:
Done: https://github.com/openshift/cluster-ingress-operator/compare/8b503220967687ec379a5e64d3a443172d2e6b70..a64df5779bcec0aeaf5c4f1e0c4e99a530c05da3